chia/full_node/full_node_api.py

```diff
@@ -119,9 +122,56 @@ async def new_transaction(self, transaction: full_node_protocol.NewTransaction)
             return None
 
         if self.full_node.mempool_manager.is_fee_enough(transaction.fees, transaction.cost):
-            request_tx = full_node_protocol.RequestTransaction(transaction.transaction_id)
-            msg = make_msg(ProtocolMessageTypes.request_transaction, request_tx)
-            return msg
+            # If there's current pending request just add this peer to the set of peers that have this tx
+            if transaction.transaction_id in self.full_node.pending_tx_request:
+                if transaction.transaction_id in self.full_node.peers_with_tx:
+                    current_set = self.full_node.peers_with_tx[transaction.transaction_id]
+                    if peer.peer_node_id in current_set:
+                        return None
+                    current_set.add(peer.peer_node_id)
+                    return None
+                else:
+                    new_set = set()
+                    new_set.add(peer.peer_node_id)
+                    self.full_node.peers_with_tx[transaction.transaction_id] = new_set
+                    return None
+
+            self.full_node.pending_tx_request[transaction.transaction_id] = peer.peer_node_id
+            new_set = set()
+            new_set.add(peer.peer_node_id)
+            self.full_node.peers_with_tx[transaction.transaction_id] = new_set
+
+            async def tx_request_and_timeout(full_node: FullNode, transaction_id):
+                try:
+                    while True:
+                        if transaction_id not in full_node.peers_with_tx:
+                            break
+                        peers_with_tx: Set = full_node.peers_with_tx[transaction_id]
+                        if len(peers_with_tx) == 0:
+                            break
+                        peer_id = peers_with_tx.pop()
+                        assert full_node.server is not None
+                        if peer_id not in full_node.server.all_connections:
+                            continue
+                        peer = full_node.server.all_connections[peer_id]
+                        request_tx = full_node_protocol.RequestTransaction(transaction.transaction_id)
+                        msg = make_msg(ProtocolMessageTypes.request_transaction, request_tx)
+                        await peer.send_message(msg)
+                        await asyncio.sleep(20)
```

Comment:
20 seconds seems like a lot. Is it possible that the mempool manger seen list has been changed and it will return false even though we already saw it?

Reply:
I think requesting transactions in this manner is risky. 20 seconds is most likely too long or too short. This loop doesn't end until we have the transaction. It may send requests indefinitely. It's also independent of other requests for transactions (iiuc). If there are 100 other requests for transactions from this same peer, 20 seconds may not be long enough. And sending another request in that case will only make it worse.